### DISCUSSION TITLE
How to detect any change made on canvas

### DISCUSSION THREAD

**Discussion by wvsr** [fabricjs/fabric.js#9321] [Q&A] [answered]
I tried this way but doesn't work for text(TextBox) modification such as making text bold,italic, changing text size etc. 

Also I think written code by me not working for adding shadow, changing opacity of object etc.

Is there any way to detect all changes I mentioned?

```ts
// detect any change and save on db
const debounceCanvasChange = debounce(() => canvasChange(), 300)

newCanvas.on('object:added', () => debounceCanvasChange())
newCanvas.on('object:modified', () => {
    console.log('modified object')
    debounceCanvasChange()
})
newCanvas.on('object:removed', () => debounceCanvasChange())
```

**Comment by asturur** [comment, marked as answer]
FabricJS offers events to detect changes outside your own code, made by user interactions.
If you write code to change text from normal to bold, you are in charge of knowing that is happening, same for opacity, users can't change opacity of objects outside your own code.

You can go as simple as firing canvas.fire('object:modified') on your own in your opacity changer function.

added, modified, removed, text:changed, will cover most of the use cases, maybe not all.
If you find events that are not covered by events please let us now by adding something on the canvas and give us instruction on how trigger the modification outside of events.

**Reply by wvsr** [reply]
canvas.fire working fine for me, thanks
